replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/AbstractPersistable.java

```diff
@@ -0,0 +1,178 @@
+/**
+ * Copyright (c) Connexta
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package com.connexta.replication.api.impl.data;
+
+import com.connexta.ion.replication.api.ReplicationPersistenceException;
+import com.connexta.replication.api.data.Persistable;
+import com.connexta.replication.api.impl.persistence.pojo.Pojo;
+import com.google.common.annotations.VisibleForTesting;
+import java.util.UUID;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+/**
+ * This class defines some data that any object to be saved in persistence should have and methods
+ * to manipulate that data.
+ *
+ * @param <P> the type of pojo associated with this persistable
+ */
+public abstract class AbstractPersistable<P extends Pojo> implements Persistable {
+  /** A string representing the type of this object. Used when generating exception or logs. */
+  protected final String type;
+
+  private String id;
+
+  protected AbstractPersistable(String type) {
+    this.type = type;
+    this.id = UUID.randomUUID().toString();
+  }
+
+  protected AbstractPersistable(String type, @Nullable String id) {
```

Comment:
Why? Shouldn't all objects we persist have a unique internal identifier? 
Our original code had that restriction so I am not changing that.

Reply:
I think we should enforce and ID. We had this issue with Replication Items relying on metadata ID as the primary ID and duplicate data unknowingly overwritting each other.

Reply:
@peterhuffer Was this a reply to Chris or a comment to the existing code? I believe the existing code does enforce the presence of a unique internal id. The first constructor is meant to be used when a new object is created and the other when we are deserializing an existing one. I just added javadoc for that and updated the subclass constructors to call this one with a `null` id since the id is restored b y the `readFrom()`

Reply:
In reply to Chris' comment.

Reply:
My comment was related to the @Nullable which I don't think we should allow. We should always have an id. If we let this be nullable we should at least call the this(type) so we generate a new one.

Reply:
It is a protected constructor used  by derived classes when deserializing to avoid initializing an id before the `readFrom()` is called which does that.